harden REST API `atSlot` against non-finalized blocks

arnetheduck opened the pull request:
* harden validator API against pre-finalized slot requests
* check `syncHorizon` when responding to validator api requests too far
from `head`
* limit state-id based requests to one epoch ahead of `head`
* put historic data bounds on block/attestation/etc validator production API, preventing them from being used with already-finalized slots
* add validator block smoke tests
* make rest test create a new genesis with the tests running roughly in
the first epoch to allow testing a few more boundary conditions

arnetheduck committed:
harden `atSlot` against non-finalized blocks

* harden validator API against pre-finalized slot requests
* check `syncHorizon` when responding to validator api requests too far
from `head`
* limit state-id based requests to one epoch ahead of `head`

etan-status commented on beacon_chain/rpc/rest_utils.nim:
```diff
@@ -39,20 +40,13 @@ proc validate(key: string, value: string): int =
-  if res.slot + uint64(2 * SLOTS_PER_EPOCH) < slot:
+  let head = node.dag.head
+
+  if slot > head.slot and not node.isSynced(head):
```
`node.isSynced` returns `false` when `head` is more than `syncHorizon` behind the wall clock, it does not take into account the requested `slot`.

The old logic allowed `slot` to be `syncHorizon` ahead of `head` even while not yet fully synced.
The new logic no longer allows even requesting a single slot ahead of `head` while sync has not completed.

Intended?

arnetheduck replied:
yes, while syncing we don't want to allow spurious requests for slots that will cause expensive empty-slots replays of the state

etan-status commented on beacon_chain/rpc/rest_utils.nim:
```diff
@@ -64,7 +58,13 @@ proc getBlockSlotId*(node: BeaconNode,
+    # head to avoid long empty slot replays (in particular a second epoch
+    # transition)
+    if stateIdent.slot.epoch + 1 > node.dag.head.slot.epoch:
+      return err("Requesting state too far ahead of head current head")
```
- There is an extra `head` in the error message.
- Logic looks alright, `slot.epoch + 1` does not seem to be overflowable.

arnetheduck replied:
actually it was overflowing because of the magic far_future_epoch handling (that at some point should be applied to `+` as well...)

etan-status commented on beacon_chain/rpc/rest_utils.nim:
```diff
@@ -84,8 +84,12 @@ proc getBlockSlotId*(node: BeaconNode,
+        max(
+          getStateField(node.dag.headState, current_justified_checkpoint).epoch,
+          node.dag.finalizedHead.slot.epoch)
+      ok(node.dag.head.atEpochStart(justifiedEpoch).toBlockSlotId().expect("not nil"))
```
Good catch!

etan-status commented on beacon_chain/rpc/rest_validator_api.nim:
```diff
@@ -606,19 +615,22 @@ proc installValidatorApiHandlers*(router: var RestRouter, node: BeaconNode) =
+      let res = slot.get()
+      if res.isErr():
+        return RestApiResponse.jsonError(Http400, InvalidSlotValueError,
+                                          $res.error())
```
NIT: Indentation of the second line w.r.t. `(`, on all of these.

arnetheduck replied:
this is true, but also a bit of a chore / best-effort thing as long as we don't have an  autoformatter.. the best way to highlight  these is through the little suggestion button which gives the right perfomance/effort ratio for fixing them (and avoids typing NIT in ALL-CAPS :))

![Screenshot from 2022-03-22 20-17-04](https://user-images.githubusercontent.com/1382986/159558626-36b52d9c-ad3f-435a-a125-73ae458e04e7.png)

etan-status commented on beacon_chain/rpc/rest_validator_api.nim:
```diff
@@ -400,6 +406,9 @@ proc installValidatorApiHandlers*(router: var RestRouter, node: BeaconNode) =
               return RestApiResponse.jsonError(Http400, InvalidSlotValueError,
                                                $res.error())
             res.get()
+        if qslot <= node.dag.finalizedHead.slot:
```
The altair check of other endpoints is inside the block fetching `qslot` -- should this additional check be located similarly, as in, first loading into `rslot`, then validating against the allowed range, then assigning to `qslot`? It doesn't matter semantically, it's more of a style / consistency thing.

arnetheduck replied:
indeed, there's a myriad of these little improvements that one _could_ make with infinite time at hand..

arnetheduck committed:
more checks and fixes

arnetheduck committed:
typo

arnetheduck committed:
check finalized attestation data response

arnetheduck committed:
add validator block tests

make rest test create a new genesis with the tests running roughly in
the first epoch to allow testing a few more boundary conditions

arnetheduck committed:
more tests, simplify epoch condition
